shaders/src/light_reflections.fs

```diff
@@ -0,0 +1,276 @@
+#undef Point2
+#undef Vector2
+#undef Vector3
+#undef Vector4
+
+// -- end "BSD 2-clause license" -------------------------------------------------------------------
+
+mat4x4 scaleMatrix(float x, float y) {
+    mat4x4 m = mat4(1.0f);
+    m[0].x = x;
+    m[1].y = y;
+    m[2].z = 1.0f;
+    m[3].w = 1.0f;
+    return m;
+}
+
+mat4x4 translateMatrix(float x, float y) {
+    mat4x4 m = mat4(1.0f);
+    m[3].xy = vec2(x, y);
+    return m;
+}
+
+// random number between 0 and 1, using interleaved gradient noise
+float ssr_random(const highp vec2 w) {
+    const vec3 m = vec3(0.06711056, 0.00583715, 52.9829189);
+    return fract(m.z * fract(dot(w, m.xy)));
+}
+
+/**
+ * Evaluates screen-space reflections, storing the color in Fr if there's a hit.
+ * r is the desired reflected vector.
+ * Returns true if there's a hit, false otherwise.
+ */
+bool evaluateScreenSpaceReflections(highp vec3 r, inout vec3 Fr) {
+    highp vec3 wsRayDirection = r;
+    highp vec3 wsRayStart = shading_position + frameUniforms.ssrBias * wsRayDirection;
+
+    // ray start/end in view space
+    highp vec4 vsRayStart = getViewFromWorldMatrix() * vec4(wsRayStart, 1.0);
+    highp vec4 vsRayDirection = getViewFromWorldMatrix() * vec4(wsRayDirection, 0.0);
+
+    highp vec3 vsOrigin = vsRayStart.xyz;
+    highp vec3 vsDirection = vsRayDirection.xyz;
+    highp float vsZThickness = frameUniforms.ssrThickness;
+    // TODO: use the actual near plane.
+    highp float nearPlaneZ = -0.1f;
+    highp float stride = 1.0;
+    // TODO: jitterFraction should be a between 0 and 1, but anything < 1 gives banding artifacts.
+    highp float jitterFraction = 1.0f;
+    // TODO: set this to the larger of the viewport dimensions.
+    highp float maxSteps = 1000.0;
+    highp float maxRayTraceDistance = frameUniforms.ssrDistance;
+
+    vec2 res = vec2(textureSize(light_structure, 0).xy);
+    highp mat4x4 projectToPixelMatrix =
+        scaleMatrix(res.x, res.y) *
+        translateMatrix(0.5f, 0.5f) *
+#if defined(TARGET_METAL_ENVIRONMENT) || defined(TARGET_VULKAN_ENVIRONMENT)
+        scaleMatrix(0.5f, -0.5f) *  // Metal and Vulkan's texture space is y-inverted
+#else
+        scaleMatrix(0.5f, 0.5f) *
+#endif
+        getClipFromViewMatrix();
+
+    // Outputs from the traceScreenSpaceRay function.
+    vec2 hitPixel;  // not currently used
+    vec3 vsHitPoint;
+
+    if (traceScreenSpaceRay(vsOrigin, vsDirection, projectToPixelMatrix, light_structure,
+            vsZThickness, nearPlaneZ, stride, jitterFraction, maxSteps,
+            maxRayTraceDistance, hitPixel, vsHitPoint)) {
+        vec2 ssrRes = vec2(textureSize(light_ssr, 0).xy);
+        vec4 reprojected = scaleMatrix(ssrRes.x, ssrRes.y) *
+                translateMatrix(0.5f, 0.5f) *
+#if defined(TARGET_METAL_ENVIRONMENT) || defined(TARGET_VULKAN_ENVIRONMENT)
+                scaleMatrix(0.5f, -0.5f) *
+#else
+                scaleMatrix(0.5f, 0.5f) *
+#endif
+                frameUniforms.ssrReprojection *
```

Comment:
This can be in a uniform. And even if not, it's almost the same computation than above.